### PR TITLE
feat: (PoC) impl wasm support for playground.

### PR DESCRIPTION
**Pull request by AcrylicShrimp**
**Description:**

This PR implements checker behaviors for `wasm` binding. By supporting wasm, implementing playground will become very easy. I also implemented PoC of playground and it works well. I'll contribute on that when this PR is ready and merged.

Due to the constraints of wasm, there is a few APIs that should not be used. Below is the list.

- `std::time::Instant`
- `mimalloc_rust` from `swc_node_base`
- All file-system related APIs.

We dropped some code to avoid use of the APIs, and this PR contains those changes. I marked this PR as draft; we have to manage them somehow before merge it.

**Related issue:**
#300
#1021 

**Task**

- [ ] #1081

### CODE REVIEW

**Comment by sunrabbit123** [comment]
@kdy1 If this issue passes, I think we should create an stc-playground repository.

File: crates/stc_utils/src/lib.rs [hunk 1]
```diff
@@ -1,8 +1,5 @@
 #![feature(never_type)]
 
-/// Use good memory allocator.
-extern crate swc_node_base;
```
Comment (sunrabbit123) [inline crates/stc_utils/src/lib.rs:4]: @kdy1 

I discovery this code is unused
So I think it's safe to delete it, what do you think?

> `swc` uses this when attempting to bind to the napi.rs.

Comment (kdy1) [reply crates/stc_utils/src/lib.rs:4]: IIRC, `swc_node_base` is about using a faster mallocator

File: crates/binding_wasm/src/lib.rs [hunk 2]
```diff
@@ -1,14 +1,146 @@
...
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct CheckAndAnnotateOutput {
+    error: String,
+}
+
+pub struct PassThroughWriter {
+    pub buf: Arc<RefCell<BufWriter<Vec<u8>>>>,
```
Comment (kdy1) [inline crates/binding_wasm/src/lib.rs:88]: This should be `Arc<Mutex<Vec<u8>>>`.

 - `BufWriter<Vec<u8>>` has nothing good over `Vec<u8>` becuase it's already in memory.
 - `Arc<RefCell<T>>` is unsound, and it's `unsafe.`

File: crates/binding_wasm/src/lib.rs [hunk 3]
```diff
@@ -1,14 +1,146 @@
...
+    fn write(&mut self, buf: &[u8]) -> std::io::Result<usize> {
+        self.buf.borrow_mut().write(buf)
+    }
+
+    fn flush(&mut self) -> std::io::Result<()> {
+        self.buf.borrow_mut().flush()
+    }
+}
+
+unsafe impl Send for PassThroughWriter {}
```
Comment (kdy1) [inline crates/binding_wasm/src/lib.rs:101]: This can be removed

**Review by kdy1 — CHANGES_REQUESTED** [review]

**Comment by sunrabbit123** [comment]
@AcrylicShrimp
Can you resolve clippy error?

**Comment by AcrylicShrimp** [comment]
I resolved all problems. Should I do rebase? @sunrabbit123

**Comment by sunrabbit123** [comment]
Resolved with [commit(7dc35c8a8a0d64703ab152e)](https://github.com/dudykr/stc/pull/1082/commits/c32f50de9806222187dc35c8a8a0d64703ab152e). Thank you.

May I ask for a review? @kdy1

**Comment by sunrabbit123** [comment]
I checked the related issue
I think a name similar to `stc_playground_wasm` would be better than `binding-wasm`.

**Comment by kdy1** [comment]
>  I think a name similar to stc_playground_wasm would be better than binding-wasm.


I agree with this. In my experience, we will want many APIs for the playground not exposed to normal wasm bindings

**Review by kdy1 — APPROVED** [review]
Thanks!

**Pull request merged**